<a href="https://colab.research.google.com/github/Srestrero/AlgorithmsUN2024II/blob/main/Lab3/srestreporo_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [26]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.200
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4400.40
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [29]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [35]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

12
5
17

### Test input output files

In [30]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [31]:
%%writefile output1.txt
12

Overwriting output1.txt


In [32]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [33]:
%%writefile output2.txt
9

Overwriting output2.txt


In [36]:
!./APlusB < input1.txt

12

In [37]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [38]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [39]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [40]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [41]:
!python sum.py

124 53
177


In [42]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [43]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [44]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [45]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

3
31 56 32
1792


### Test 1: Small numbers

In [46]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Overwriting input0.txt


In [47]:
%%writefile output0.txt
0

Overwriting output0.txt


In [48]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [49]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [50]:
!./max_pairwise_product_full < input1.txt

90


In [51]:
%%writefile output1.txt
90

Overwriting output1.txt


In [52]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [53]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [54]:
%%writefile output2.txt
90

Overwriting output2.txt


In [55]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [56]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [57]:
200000*200000

40000000000

In [58]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [59]:
!./max_pairwise_product_full < input3.txt

1345294336


In [60]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [61]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [62]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [63]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [64]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [65]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [66]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [67]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [68]:
%%writefile input4.txt
12
1 700 4567892 32 12 4 90 3 200000 9 200000 23

Writing input4.txt


In [ ]:
4567892 * 200000

913578400000

In [69]:
%%writefile output4.txt
913578400000

Writing output4.txt


In [70]:
%%writefile input5.txt
15
1 800 5678901 45 13 5 100 4 300000 10 300000 24 6000000 7000000 8

Writing input5.txt


In [ ]:
7000000 * 6000000

42000000000000

In [71]:
%%writefile output5.txt
42000000000000

Writing output5.txt


In [72]:
%%writefile input6.txt
18
1 1000 8901234 50 15 6 150 5 500000 12 500000 30 9000000 10000000 9 12000000 15000000 7

Writing input6.txt


In [ ]:
15000000 * 12000000

180000000000000

In [73]:
%%writefile output6.txt
180000000000000

Writing output6.txt


In [74]:
%%writefile input7.txt
20
2 2000 12345678 60 18 8 200 6 800000 15 800000 35 20000000 25000000 10 30000000 40000000 9 45000000 50000000

Writing input7.txt


In [ ]:
50000000 * 45000000

2250000000000000

In [75]:
%%writefile output7.txt
2250000000000000

Writing output7.txt


In [76]:
%%writefile input8.txt
22
3 3000 23456789 70 20 10 300 7 1000000 18 1000000 40 50000000 60000000 12 70000000 80000000 11 90000000 100000000 110000000 120000000

Writing input8.txt


In [ ]:
120000000 * 110000000

13200000000000000

In [77]:
%%writefile output8.txt
13200000000000000

Writing output8.txt


In [78]:
%%writefile input9.txt
25
4 4000 34567890 80 25 12 400 8 2000000 20 2000000 45 150000000 160000000 15 170000000 180000000 13 190000000 200000000 210000000 220000000 230000000 240000000 250000000

Writing input9.txt


In [ ]:
250000000 * 240000000

60000000000000000

In [79]:
%%writefile output9.txt
60000000000000000

Writing output9.txt


In [80]:
%%writefile input10.txt
30
5 5000 45678901 100 30 15 500 10 3000000 25 3000000 50 300000000 350000000 18 400000000 450000000 16 500000000 600000000 650000000 700000000 750000000 800000000 850000000 900000000 950000000 999999999 1000000000 1000000000

Writing input10.txt


In [ ]:
1000000000 * 1000000000

1000000000000000000

In [81]:
%%writefile output10.txt
1000000000000000000

Writing output10.txt


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [82]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [83]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [84]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [85]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [86]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [87]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

2147483646

real	5m7.549s
user	5m4.787s
sys	0m0.095s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code




In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

^C


In [ ]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

^C


In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

/bin/bash: line 1: input13.txt: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


## Python

In [7]:
%%writefile max_pairwise_product_execints.py
def MaxPairwiseProduct(numbers, execints):                      # INSTRUCTIONS executed by line
    max_product = 0                                             # 1
    execints[0] += 1

    aux_product = 0                                             # 1
    execints[0] += 1

    n = len(numbers)                                            # 1
    execints[0] += 1

    execints[0] += 1
    for first in range(n):                                      # 2 + 2n
        execints[0] += 2

        execints[0] += 1
        for second in range(n):                                 # 2n + 2n**2
            execints[0] += 2

            execints[0] += 1
            if first != second:                                 # n**2
                aux_product = numbers[first] * numbers[second]   # n**2 - n
                execints[0] += 1

                execints[0] += 1
                if aux_product > max_product:                    # n**2 - n
                    max_product = aux_product                    # [0, n**2/2 - n/2] upper bound
                    execints[0] += 1

        execints[0] += 1
    execints[0] += 1

    return max_product                                          # 5 + 2n + [5n**2, 11n**2/2 - n/2] upper bound

def main():                                                     # INSTRUCTIONS executed by line
    execints = [0]                                             # Usando lista para simular paso por referencia

    n = int(input())                                           # 1
    execints[0] += 1

    numbers = list(map(int, input().split()))                  # n
    execints[0] += n

    execints[0] += 1
    result = MaxPairwiseProduct(numbers, execints)             # 1 + 5 + 2n + [5n**2, 11n**2/2 - n/2]

    execints[0] += 1
    print(execints[0])                                         # Imprime el número total de instrucciones ejecutadas

    return 0                                                   # --------------
                                                              # 11 + 6n + [5n**2, 11n**2/2 - n/2] upper bound

if __name__ == '__main__':
    main()

Overwriting max_pairwise_product_execints.py


La complejidad temporal sigue siendo la misma:
$$T(n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2} - \frac{n}{2}\right) > 6n^2 = O(n^2)$$

In [13]:
# Verify
!python3 max_pairwise_product_execints.py < input0.txt

538


In [17]:
!python3 max_pairwise_product_execints.py < input2.txt

539


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [18]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [19]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 7n + \left[2(n-1)n+1, 2.5(n-1)n\right)= O(n^2)$$

Then for the input0.txt
$$T = 11 + 7(10) + 2(9)11   = 279 $$

### Estimate and verify the maximun n that can be solved in 1 second



Analysis of Maximum n that can be solved in 1 second

First, let's look at the time complexity from the C++ code:
$$T(n) = 11 + 7n + [2(n-1)n+1, 2.5(n-1)n]$$
Let's expand the formula:

Lower bound: $2(n-1)n + 1 = 2n^2 - 2n + 1$
Upper bound: $2.5(n-1)n = 2.5n^2 - 2.5n$



Therefore, the complete range of instructions is:
$$T(n) = 11 + 7n + [2n^2 - 2n + 1, 2.5n^2 - 2.5n]$$

Assuming a modern computer can execute approximately $10^8$ operations per second, we can set up the inequality using the lower bound:

$$11 + 7n + (2n^2 - 2n + 1) \leq 10^8$$
Simplifying:
$$2n^2 + 5n + 12 \leq 10^8$$

Since the quadratic term dominates, we can approximate:
$$2n^2 \leq 10^8$$
$$n^2 \leq 5 \times 10^7$$
$$n \leq \sqrt{5 \times 10^7} \approx 7071$$

In [100]:
%%writefile test_generator.cpp
#include <iostream>
#include <fstream>
#include <random>
#include <string>

void generateTest(int n, const std::string& filename) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(1, 1000000000);  // Números entre 1 y 10^9

    std::ofstream file(filename);
    file << n << "\n";
    for (int i = 0; i < n; i++) {
        file << dis(gen);
        if (i < n-1) file << " ";
    }
    file << "\n";
    file.close();
}

int main() {
    // Generar tests con tamaños crecientes
    int sizes[] = {1000, 2000, 3000, 4000, 5000, 6000, 7000};

    for (int size : sizes) {
        std::string filename = "input_" + std::to_string(size) + ".txt";
        generateTest(size, filename);
        std::cout << "Generated " << filename << " with " << size << " numbers\n";
    }

    return 0;
}

Overwriting test_generator.cpp


In [102]:
!g++ test_generator.cpp -o test_generator
!./test_generator

# Compilar el programa principal
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long

# Ejecutar los tests y medir el tiempo
!time ./max_pairwise_product_long < input_1000.txt
!time ./max_pairwise_product_long < input_2000.txt
!time ./max_pairwise_product_long < input_3000.txt
!time ./max_pairwise_product_long < input_4000.txt
!time ./max_pairwise_product_long < input_5000.txt
!time ./max_pairwise_product_long < input_6000.txt
!time ./max_pairwise_product_long < input_7000.txt

Generated input_1000.txt with 1000 numbers
Generated input_2000.txt with 2000 numbers
Generated input_3000.txt with 3000 numbers
Generated input_4000.txt with 4000 numbers
Generated input_5000.txt with 5000 numbers
Generated input_6000.txt with 6000 numbers
Generated input_7000.txt with 7000 numbers
Generated input_8000.txt with 8000 numbers
998332035126414263

real	0m0.008s
user	0m0.006s
sys	0m0.002s
999026053556593244

real	0m0.018s
user	0m0.017s
sys	0m0.001s
998866083570378509

real	0m0.034s
user	0m0.032s
sys	0m0.001s
999165755656685760

real	0m0.057s
user	0m0.056s
sys	0m0.000s
998956441128546623

real	0m0.093s
user	0m0.089s
sys	0m0.003s
999056568192190536

real	0m0.122s
user	0m0.119s
sys	0m0.003s
999563069190265523

real	0m0.171s
user	0m0.165s
sys	0m0.005s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Timsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Test with inputs 1 and 2



Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3


Calculate estimated time by instruction and number of instructions by second

### Estimate and verify the maximun n that can be solved in 1 second

## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

Test with inputs 1 and 2



Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3


Calculate estimated time by instruction and number of instructions by second

### Estimate and verify the maximun n that can be solved in 1 second

## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


Test with inputs 1 and 2




Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

Test with inputs 1 and 2




Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second

### Estimate and verify the maximun n that can be solved in 1 second

## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".